src/lib.rs

```diff
@@ -99,17 +102,61 @@ impl Diagnostic {
     }
 }
 
-pub type Sources<'a> = [&'a dyn SourceContainer];
+pub type Sources<'a> = [&'a mut dyn SourceContainer];
 
 /// SourceContainers offer source-code to be compiled via the load_source function.
 /// Furthermore it offers a location-String used when reporting diagnostics.
 pub trait SourceContainer {
     /// loads and returns the SourceEntry that contains the SourceCode and the path it was loaded from
-    fn load_source(&self) -> Result<SourceCode, String>;
+    fn load_source(&mut self, encoding: Option<&'static Encoding>) -> Result<SourceCode, String>;
     /// returns the location of this source-container. Used when reporting diagnostics.
     fn get_location(&self) -> &str;
 }
 
+pub struct FilePath {
+    pub path: String,
+}
+
+impl<T> SourceContainer for T
+where
+    T: Read,
+{
+    fn load_source(&mut self, encoding: Option<&'static Encoding>) -> Result<SourceCode, String> {
+        let mut buffer = String::new();
+        let mut decoder = DecodeReaderBytesBuilder::new()
+            .encoding(encoding)
+            .build(self);
+        decoder
+            .read_to_string(&mut buffer)
+            .map_err(|err| format!("{:}", err))?;
+        Ok(SourceCode {
+            source: buffer,
+            path: "local".into(),
```

Comment:
Is this only used in tests? This may cause confusion if it ever appears in the production mode

Reply:
Yes, but I wanted to have the actual reading of the strings shared with productive code, so I added "local" for this one. I am not a fan honestly.